packages/unplugin-typegpu/src/babel.ts

```diff
@@ -17,15 +17,92 @@ import {
 const template = (
   Babel as unknown as { packages: { template: typeof TemplateGenerator } }
 ).packages.template;
-const types = (Babel as unknown as { packages: { types: typeof t } }).packages
-  .types;
+const types = (Babel as unknown as { packages: { types: typeof babel } })
+  .packages.types;
+
+function isKernelMarkedFunction(
+  node:
+    | babel.FunctionDeclaration
+    | babel.FunctionExpression
+    | babel.ArrowFunctionExpression,
+) {
+  const directives = (
+    'directives' in node.body ? (node.body?.directives ?? []) : []
+  ).map((directive) => directive.value.value);
+
+  return directives.includes('kernel');
+}
+
+function functionToTranspiled(
+  node: babel.ArrowFunctionExpression | babel.FunctionExpression,
+  ctx: Context,
+): babel.CallExpression | null {
+  if (!isKernelMarkedFunction(node)) {
+    return null;
+  }
+
+  const { argNames, body, externalNames } = transpileFn(node);
+  const tgpuAlias = ctx.tgpuAliases.values().next().value;
+  if (tgpuAlias === undefined) {
+    throw new Error(
+      `No tgpu import found, cannot assign ast to function in file: ${ctx.fileId ?? ''}`,
+    );
+  }
+
+  return types.callExpression(
+    template.expression(`${tgpuAlias}.__assignAst`)(),
+    [
+      node,
+      template.expression`${embedJSON({ argNames, body, externalNames })}`(),
+      types.objectExpression(
+        externalNames.map((name) =>
+          types.objectProperty(types.identifier(name), types.identifier(name)),
+        ),
+      ),
+    ],
+  );
+}
 
 function functionVisitor(ctx: Context): TraverseOptions {
   return {
     ImportDeclaration(path) {
       gatherTgpuAliases(path.node, ctx);
     },
 
+    ArrowFunctionExpression(path) {
+      const transpiled = functionToTranspiled(path.node, ctx);
+      if (transpiled) {
+        path.replaceWith(transpiled);
+        path.skip();
+      }
+    },
+
+    FunctionExpression(path) {
+      const transpiled = functionToTranspiled(path.node, ctx);
+      if (transpiled) {
+        path.replaceWith(transpiled);
+        path.skip();
+      }
+    },
+
+    FunctionDeclaration(path) {
+      const node = path.node;
+      const expression = types.functionExpression(
+        node.id,
+        node.params,
+        node.body,
+      );
+      const transpiled = functionToTranspiled(expression, ctx);
+      if (transpiled && node.id) {
+        path.replaceWith(
+          types.variableDeclaration('const', [
```

Comment:
function statements are no longer hoisted after this replacement, but I don't see how else to do this